nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py

```diff
@@ -0,0 +1,418 @@
+            ).mean()
+        else:
+
+            loss_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            kl_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+
+            rewards_chosen_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_rejected_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_std = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_margins = torch.tensor(0.0, device=torch.cuda.current_device())
+
+        # we can only log on one rank if it is rank zero so we broadcast from last rank
+        torch.distributed.broadcast(loss_mean, get_last_rank())
+        torch.distributed.broadcast(kl_mean, get_last_rank())
+
+        torch.distributed.broadcast(rewards_chosen_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_rejected_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_std, get_last_rank())
+        torch.distributed.broadcast(rewards_margins, get_last_rank())
+
+        metrics = {
+            "loss": loss_mean,
+            "kl": kl_mean,
+            "rewards_chosen_mean": rewards_chosen_mean,
+            "rewards_rejected_mean": rewards_rejected_mean,
+            "rewards_all_mean": rewards_all_mean,
+            "rewards_all_std": rewards_all_std,
+            "rewards_margin": rewards_margins,
+        }
+
+        # move to CPU
+        metrics = {k: v.item() for k, v in metrics.items()}
+
+        return loss_mean.item(), metrics
+
+    def prepare_for_training_step(self):
+        # custom trainers will always zero grad for us
+        prepare_for_training_step(self, zero_grad=False)
+
+    def finish_training_step(self):
+        grad_reductions(self)
+
+    def prepare_for_validation_step(self):
+        prepare_for_validation_step(self)
+
+    def finish_validation_step(self):
+        finish_validation_step(self)
+
+    @torch.no_grad()
+    def get_logprob_batch(self, batch):
+        seq_length = batch["samples"].shape[1]
+        batch_size = batch["samples"].shape[0]
+
+        # num_microbatches = get_num_microbatches()
+        num_microbatches = divide(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size)
+        data_iter = get_iterator_k_split(batch, num_microbatches)
+        set_sync_funcs(self, forward_only=True)
+
+        fwd_bwd_function = get_forward_backward_func()
+
+        logprobs_list = fwd_bwd_function(
+            forward_step_func=self.get_forward_output_and_loss_func(logprobs_only=True),
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=num_microbatches,
+            forward_only=True,
+            seq_length=seq_length,
+            micro_batch_size=self.cfg.kto.log_prob_forward_micro_batch_size,
```

Comment:
Should be: `micro_batch_size=min(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size),`

Just a guard in case someone enters a value too high for the forward mbs

Reply:
Good idea.